pkg/manager/member/tiflash_scaler.go

```diff
@@ -47,13 +53,114 @@ func (tfs *tiflashScaler) Scale(tc *v1alpha1.TidbCluster, oldSet *apps.StatefulS
 }
 
 func (tfs *tiflashScaler) ScaleOut(tc *v1alpha1.TidbCluster, oldSet *apps.StatefulSet, newSet *apps.StatefulSet) error {
+	_, ordinal, replicas, deleteSlots := scaleOne(oldSet, newSet)
+	resetReplicas(newSet, oldSet)
+	if tc.TiFlashUpgrading() {
+		return nil
+	}
+
+	klog.Infof("scaling out tiflash statefulset %s/%s, ordinal: %d (replicas: %d, delete slots: %v)", oldSet.Namespace, oldSet.Name, ordinal, replicas, deleteSlots.List())
+	_, err := tfs.deleteDeferDeletingPVC(tc, oldSet.GetName(), v1alpha1.TiFlashMemberType, ordinal)
+	if err != nil {
+		return err
+	}
 
+	setReplicasAndDeleteSlots(newSet, replicas, deleteSlots)
 	return nil
 }
 
 func (tfs *tiflashScaler) ScaleIn(tc *v1alpha1.TidbCluster, oldSet *apps.StatefulSet, newSet *apps.StatefulSet) error {
+	ns := tc.GetNamespace()
+	tcName := tc.GetName()
+	// we can only remove one member at a time when scaling in
+	_, ordinal, replicas, deleteSlots := scaleOne(oldSet, newSet)
+	resetReplicas(newSet, oldSet)
+
+	// tiflash can not scale in when it is upgrading
+	if tc.TiFlashUpgrading() {
+		klog.Infof("TidbCluster: [%s/%s]'s tiflash is upgrading, postpone the scale in until the upgrade completes",
+			ns, tcName)
+		return nil
+	}
 
-	return nil
+	klog.Infof("scaling in tiflash statefulset %s/%s, ordinal: %d (replicas: %d, delete slots: %v)", oldSet.Namespace, oldSet.Name, ordinal, replicas, deleteSlots.List())
+	// We need delete store from cluster before decreasing the statefulset replicas
+	podName := ordinalPodName(v1alpha1.TiFlashMemberType, tcName, ordinal)
+	pod, err := tfs.podLister.Pods(ns).Get(podName)
+	if err != nil {
+		return err
+	}
+
+	if controller.PodWebhookEnabled {
```

Comment:
I think we can manage tiflash scaling in the webhook  instead of in the controller as webhook would be the default choice to manage the lifecyle of each component?

Reply:
When will Webhook be the default choice? I think it's not decided yet, btw, before v1.15, all of the requests for Pod will be sent to the Webhook if enabled, if something wrong with the Webhook, it will block the whole cluster, so I would like to support it in controller first.